Palette.py - tool for importing palette files of various types

Softdrink117 opened the pull request:
Heya, my apologies for the extremely long delay. Following up on the conversation from [this issue](https://github.com/DragonMinded/libnaomi/issues/4) from May 2023.

I added support for several new file types; the tool now works with `.pal` and `.gpl` palettes, Adobe `.ase` and `.aco` swatch files, `.png` images, and Paint.Net `.txt` palettes. 

I also managed to get proper integration in the build chain working with a couple changes to the `Makefile.shared` file.

I've created a sample based on the existing examples (`spritetest`, `carttest`, etc.), which is located [here](https://github.com/Softdrink117/naomi-samples/tree/main/palettetest) for your review; wasn't sure if it was appropriate to include it directly in this PR, but happy to add it if you feel like it's a good fit.

Softdrink117 committed:
Add palette.py to tools

Softdrink117 committed:
Working make integration, palette.py bugfix

Softdrink117 committed:
Minor cleanup

Softdrink117 committed:
Add txt to PALOBJS in Makefile.shared

Softdrink117 committed:
Merge pull request #1 from Softdrink117/20230902

Palette.py working

DragonMinded commented:
I think the example would be fine for inclusion. It gives working code that also shows what can be achieved and how to add the makefile entries needed. I've tried to keep comprehensive examples as the documentation isn't great in this project (yet?), so outside of working examples discoverability is a bit limited. If you want, go ahead and include it in this PR directory (don't forget to update the places that compile all the examples and such).

DragonMinded commented on tools/palette.py:
```diff
@@ -0,0 +1,645 @@
+    palette_type = PaletteType.Unknown
+    args = None
+
+    def __init__(self,ar):
```
In general, I'd prefer fully typed files, but I don't think I've run mypy on the rest of them to verify either. So, ¯\\\_(ツ)\_/¯

Softdrink117 replied:
Ah, got it. I actually very rarely use Python, so I'm honestly not as familiar with static typing in this language. Dynamic typing and type ambiguity are really not my preference either, so I'll go through again and see if I can disambiguate any further.

DragonMinded commented on tools/palette.py:
```diff
@@ -0,0 +1,645 @@
+        output.color_count = len(output.palette_colors)
+
+    # Validate color count is either 16 or 256
+    if output.color_count != 16 and output.color_count != 256:
+        raise Exception(f"ERROR: Parsed color count does not align with supported NAOMI palette sizes! {output.color_count} != 16 or 256")
```
I wonder if it makes sense here to just fill the rest with an obnoxious color like (0x255, 0, 0x255). Presumably, somebody converting a palette that doesn't have all 16/256 colors is working with textures that simply do not have that many colors in them.

DragonMinded commented on tools/palette.py:
```diff
@@ -0,0 +1,645 @@
+    # Determine the number of pixels to encode as a palette
+    # Automatically take the larger possible size if possible
+    pix_count = width * height
+    if pix_count < 16:
+        raise Exception(f"ERROR: PNG pixel count is smaller than minimum NAOMI palette size! {width * height} < 16")
+    elif pix_count >= 16 and pix_count < 256:
+        pix_count = 16
+    else:
+        pix_count = 256
```
Same comment here, it seems we should probably just size the palette up to the next increment (16 or 256) and only throw if its larger than 256.

Softdrink117 replied:
Sure thing, can do. I was actually waffling back and forth between rejecting or resizing off-size palettes when I was implementing this; both solutions have some downsides, but I think you're right that upsizing (and maybe throwing a warning if it upsizes from 16 -> 256, since that's a pretty significant difference in terms of actual use in a project) is the friendlier option.

Softdrink117 committed:
Update palette.py

Softdrink117 committed:
Add palettetest to examples

Softdrink117 committed:
Merge pull request #2 from Softdrink117/20230905_Palette_Updates

20230905 palette updates per PR feedback

Softdrink117 committed:
Fixed type hint on __str__

Softdrink117 committed:
Merge pull request #3 from Softdrink117/20230905_Palette_Updates

Fixed type hint on __str__

Softdrink117 commented:
OK, I've made the suggested changes:

- Added type hints to a few functions that were missing them.
- Created a new function for validation and palette resizing (`validate_and_resize_palette()`).
- `palette.py` now resizes off-size palettes instead of rejecting them outright.
- There are two resize modes, `EXPAND` and `TRIM`, with `EXPAND` being the default. Expand mode steps up to the next available size, including from 16 to 256 colors. Trim mode will expand palettes that are less than 16 colors, but will truncate palettes that are between 16 and 256, or larger than 256 colors.
- Added `palettetest` to the examples, and updated the example Makefile to accommodate it.

DragonMinded commented on examples/palettetest/Makefile:
```diff
@@ -0,0 +1,90 @@
+# Please see the README for setting up a valid build environment.
+
+# Commands
+MKDIR 		:= mkdir
```
Not to split hairs, but I don't see this used anywhere?

DragonMinded replied:
Okay well I am exactly splitting hairs lmao, sorry. Anyway, am I missing where this is used?

Softdrink117 replied:
Lmao yeah I just left this in by accident; the palette test spun off a different project and I must've copy-pasted that over by accident. I'll remove it in a quick revision. Apologies!

DragonMinded reviewed (approved):
This seems to work fine locally when I tried it. Aside from the MKDIR thing, the only thing that looks missing is you didn't run "make copy" to update the binaries/ directory (how would you have known, though). I usually do that to keep example binaries updated for people to try by just sticking on their Naomi or in Demul. The only thing you should see for the rest of the binaries is an update to the header for the build date, but there should be a new file for palettetest.bin (which I saw show up locally).

Softdrink117 committed:
`make copy`, clean up palettetest Makefile

Softdrink117 committed:
Merge pull request #4 from Softdrink117/20230905_Palette_Updates

`make copy`, clean up palettetest Makefile

Softdrink117 commented:
OK, cleaned up the palettetest Makefile (removed MKDIR and reformatted to match some of the other example Makefiles more closely) and ran `make copy`!

While almost all of the changes to `/binaries` went as expected, I do see a diff of 72KB on spritetest.bin. Not sure if that is at all significant.

DragonMinded commented:
When I check out this PR, and then diff the files against the main branch after building locally, I get the expected changes (nothing different at all except for the dates, and addition of the new palette test), but when I look at your changes, there's a ton. I suspect that your repo is "dirty" compared to main, aka this branch was merged into from something you're working on where you've modified libnaomi core. I'm gonna just merge this and then recompile the binaries and call it good instead of forcing you to undo everything.

Softdrink117 commented:
Hmm, I thought that I had switched back to the unmodified core before compiling these.

Incredibly sorry for the inconvenience!
